e3nn/rsh.py

```diff
@@ -0,0 +1,167 @@
+# pylint: disable=not-callable, no-member, invalid-name, line-too-long, unexpected-keyword-arg, too-many-lines
+"""
+Real Spherical Harmonics equivariant with respect to o3.rot and o3.irr_repr
+"""
+import math
+
+import torch
+from scipy import special
+
+from e3nn.util.default_dtype import torch_default_dtype
+
+
+def spherical_harmonics_expand_matrix(lmax):
+    """
+    :return: tensor [l, m, l * m]
+    """
+    m = torch.zeros(lmax + 1, 2 * lmax + 1, sum(2 * l + 1 for l in range(lmax + 1)))
+    i = 0
+    for l in range(lmax + 1):
+        m[l, lmax - l: lmax + l + 1, i:i + 2 * l + 1] = torch.eye(2 * l + 1)
+        i += 2 * l + 1
+    return m
+
+
+def _legendre(l, z):
+    """
+    associated Legendre polynomials
+
+    :param l: int
+    :param z: tensor of shape [...]
+    :return: tensor of shape [..., m]
+    """
+    if not z.requires_grad:
+        return torch.stack([(-1)**m * torch.tensor(special.lpmv(m, l, z.cpu().double().numpy())).to(z) for m in range(-l, l + 1)], dim=-1)
+
+    fac = math.factorial(l)
+    sqz2 = (1 - z ** 2) ** 0.5
+    hsqz2 = 0.5 * sqz2
+    ihsqz2 = z / hsqz2
+
+    if l == 0:
+        return z.new_ones(*z.size(), 1)
+    if l == 1:
+        return torch.stack([-0.5 * sqz2, z, sqz2], dim=-1)
+
+    plm = [(1 - 2 * abs(l - 2 * l // 2)) * hsqz2 ** l / fac]
+    plm.append(-plm[0] * l * ihsqz2)
+    for mr in range(1, l):
+        plm.append((mr - l) * ihsqz2 * plm[mr] - (2 * l - mr + 1) * mr * plm[mr - 1])
```

Comment:
Maybe I'm being ambitious, but should we do this in cython?

Reply:
I don't want to use cython, it will complexify the installation process again. This code is used only in the case of `requires_grad=True` so I don't care much.